{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c24a89d9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting position of RBD in GISAID ref genome (with gaps): 22567\n",
      "Ending position of RBD in GISAID ref genome (with gaps): 23235\n"
     ]
    }
   ],
   "source": [
    "### Open Reading Frame for ncbi reference sequence RBD from https://www.ncbi.nlm.nih.gov/gene/1489668:\n",
    "\n",
    "################################################\n",
    "# Start sequence of RBD in NCBI: AGGGTTGTTCCCTCA\n",
    "# End sequence of RBD in NCBI: CCAGTGTGTCAATTTT\n",
    "\n",
    "# FROM MSA between GISAID ref genome and NCBI ref genome \n",
    "# Start sequence of RBD in GISAID: AGAGTCCAACCAACAG\n",
    "# End sequence of RBD in GISAID: CAAATGTGTCAATTTC\n",
    "################################################\n",
    "\n",
    "# NOTE: Since ORF indices for RBD belongs to NCBI ref genome, it needs to be converted to that for \n",
    "# GISAID ref genome since MSA file is based on that. So, to do that, the two ref genomes were aligned \n",
    "# using MUSCLE: https://www.ebi.ac.uk/Tools/msa/muscle/\n",
    "# With this MSA, and start and end sites for RBD being 22407 and 23072 in NCBI reference genome, \n",
    "# the position of these two sites were obtained from the MSA file below:\n",
    "\n",
    "def find_positions_RBD_in_GISAID_msa(msa_fasta_file):\n",
    "    with open(msa_fasta_file, \"r\") as file:\n",
    "        lines = file.readlines()\n",
    "\n",
    "    # Combine the lines to form a single sequence.\n",
    "    sequence = \"\".join([line.strip() for line in lines[1:]])  # Skip the header line.\n",
    "\n",
    "    # Initialize counters for non-gap and total characters, and positions.\n",
    "    non_gap_count = 0\n",
    "    total_count = 0\n",
    "    position_RBD_start = None\n",
    "    position_RBD_end = None\n",
    "\n",
    "    # Iterate through the sequence to find the positions and count characters.\n",
    "    for i, char in enumerate(sequence):\n",
    "        total_count += 1\n",
    "        if char != \"-\":\n",
    "            non_gap_count += 1\n",
    "            if non_gap_count == 22407:\n",
    "                position_RBD_start = i + 1  # Adjust for 1-based indexing.\n",
    "            if non_gap_count == 23072:\n",
    "                position_RBD_end = i + 1\n",
    "\n",
    "    if position_RBD_start is not None and position_RBD_end is not None:\n",
    "        print(\"Starting position of RBD in GISAID ref genome (with gaps):\", position_RBD_start)\n",
    "        print(\"Ending position of RBD in GISAID ref genome (with gaps):\", position_RBD_end)\n",
    "\n",
    "def main():\n",
    "    msa_fasta_file = \"your/path/MUSCLE_MSA_refGenomes.fasta\"  # Change this to the actual file path.\n",
    "    find_positions_RBD_in_GISAID_msa(msa_fasta_file)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "10e8f581",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Position of RBD start site in MSA between NCBI and GISAID (including gaps): 22567\n",
      "Position of RBD stop site in MSA between NCBI and GISAID (including gaps): 23235\n",
      "Position of RBD start site in GISAID  without gaps: 22517\n",
      "Position of RBD stop site in GISAID without gaps: 23185\n"
     ]
    }
   ],
   "source": [
    "## After the start and stop sites for RBD in MSA between the two ref genomes were obtained, the gap characters \n",
    "## were removed from the GISAID ref genome to identify the position of the start and stop sites of RBD below:\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "def count_and_find_positions(msa_file):\n",
    "    # Read the MSA FASTA file\n",
    "    records = list(SeqIO.parse(msa_file, \"fasta\"))\n",
    "\n",
    "    # Check if the file has at least two sequences\n",
    "    if len(records) < 2:\n",
    "        print(\"Error: MSA file should contain at least two sequences.\")\n",
    "        return\n",
    "\n",
    "    # Extract the second sequence\n",
    "    sequence = str(records[1].seq)\n",
    "\n",
    "    # Count characters at positions 22567 and 23235 (including gaps)\n",
    "    count_at_22567 = sequence.count('-', 0, 22566) + sequence[0:22567].count('A') + sequence[0:22567].count('T') + sequence[0:22567].count('G') + sequence[0:22567].count('C')\n",
    "    count_at_23235 = sequence.count('-', 0, 23234) + sequence[0:23235].count('A') + sequence[0:23235].count('T') + sequence[0:23235].count('G') + sequence[0:23235].count('C')\n",
    "\n",
    "    print(f\"Position of RBD start site in MSA between NCBI and GISAID (including gaps): {count_at_22567}\")\n",
    "    print(f\"Position of RBD stop site in MSA between NCBI and GISAID (including gaps): {count_at_23235}\")\n",
    "\n",
    "    # Find the positions of characters without gaps\n",
    "    pos_22567_no_gaps = count_at_22567 - sequence[0:22567].count('-')\n",
    "    pos_23235_no_gaps = count_at_23235 - sequence[0:23235].count('-')\n",
    "\n",
    "    print(f\"Position of RBD start site in GISAID  without gaps: {pos_22567_no_gaps}\")\n",
    "    print(f\"Position of RBD stop site in GISAID without gaps: {pos_23235_no_gaps}\")\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    # Provide the path to your MSA FASTA file\n",
    "    msa_file_path = \"your/path/MUSCLE_MSA_refGenomes.fasta\"\n",
    "    \n",
    "    count_and_find_positions(msa_file_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae8ef9b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "## RBD extraction from GISAID MSA file here onwards"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e54afe89",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting position of RBD in GISAID ref genome (with gaps): 35011\n",
      "Ending position of RBD in GISAID ref genome (with gaps): 36610\n"
     ]
    }
   ],
   "source": [
    "# Python script identifies the start and end sites for RBD in reference genome when gaps are added for the\n",
    "# GISAID MSA sequences by counting nucleotides (A,T,G or C) only and ignoring gap character in \"-\" and then \n",
    "# counting these positions after gaps are added to get the actual location for these two \n",
    "# positions when gap is present\n",
    "\n",
    "### Open Reading Frame for RBD from https://www.ncbi.nlm.nih.gov/gene/1489668:\n",
    "\n",
    "##########################################################\n",
    "\n",
    "## Start position of RBD in GISAID reference genome: 22517\n",
    "## End position of RBD in GISAID reference genome: 23187 (2 NT ADDED TO MAKE MULTIPLE OF 3)\n",
    "\n",
    "def find_positions_RBD_in_msa_fasta(msa_fasta_file):\n",
    "    with open(msa_fasta_file, \"r\") as file:\n",
    "        lines = file.readlines()\n",
    "\n",
    "    # Combine the lines to form a single sequence.\n",
    "    sequence = \"\".join([line.strip() for line in lines[1:]])  # Skip the header line.\n",
    "\n",
    "    # Initialize counters for non-gap and total characters, and positions.\n",
    "    non_gap_count = 0\n",
    "    total_count = 0\n",
    "    position_RBD_start = None\n",
    "    position_RBD_end = None\n",
    "\n",
    "    # Iterate through the sequence to find the positions and count characters.\n",
    "    for i, char in enumerate(sequence):\n",
    "        total_count += 1\n",
    "        if char != \"-\":\n",
    "            non_gap_count += 1\n",
    "            if non_gap_count == 22517:\n",
    "                position_RBD_start = i + 1  # Adjust for 1-based indexing.\n",
    "            if non_gap_count == 23187:\n",
    "                position_RBD_end = i + 1\n",
    "\n",
    "    if position_RBD_start is not None and position_RBD_end is not None:\n",
    "        print(\"Starting position of RBD in GISAID ref genome (with gaps):\", position_RBD_start)\n",
    "        print(\"Ending position of RBD in GISAID ref genome (with gaps):\", position_RBD_end)\n",
    "\n",
    "def main():\n",
    "    msa_fasta_file = \"your/path/ref_genome_MSA.fasta\"  # Change this to the actual file path.\n",
    "    find_positions_RBD_in_msa_fasta(msa_fasta_file)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7e2c078c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing sequences: 3321710it [2:15:18, 409.14it/s]                             "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Trimmed sequences between positions 35011 and 36610 to '/mmfs1/projects/changhui.yan/DeewanB/gisaid_data/main_MSA_files/trimmed_RBD.fasta'.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# Python script to cleave the rest of the MSA sequences at the position corresponding \n",
    "# to spike RBD sequence in reference sequence after adding the gaps:\n",
    "\n",
    "########################################################################\n",
    "# RBD Start Position (22517 in ref genome) in MSA with gaps: 35011           \n",
    "# RBD End Ending Position (23187 in ref genome) in MSA with gaps: 36610\n",
    "\n",
    "######################### WORKING SCRIPT ###############################\n",
    "\n",
    "import torch\n",
    "from tqdm import tqdm\n",
    "\n",
    "def trim_and_write_sequences(input_fasta, output_fasta, start_position, end_position):\n",
    "    # Count the number of sequences for progress tracking\n",
    "    with open(input_fasta, \"r\") as infile:\n",
    "        num_sequences = sum(1 for line in infile if line.startswith(\">\"))\n",
    "\n",
    "    with open(input_fasta, \"r\") as infile, open(output_fasta, \"w\") as outfile:\n",
    "        current_sequence = \"\"\n",
    "        current_header = \"\"\n",
    "\n",
    "        for line in tqdm(infile, total=num_sequences, desc=\"Processing sequences\"):\n",
    "            if line.startswith(\">\"):\n",
    "                # Write the previous sequence if any\n",
    "                if current_sequence:\n",
    "                    # Convert the sequence to a tensor\n",
    "                    sequence_tensor = torch.tensor([ord(c) for c in current_sequence], dtype=torch.int32, device='cuda')\n",
    "                    # Trim and replace \"-\" with \"\"\n",
    "                    trimmed_sequence_tensor = sequence_tensor[start_position - 1:end_position]\n",
    "                    trimmed_sequence = ''.join(chr(c) for c in trimmed_sequence_tensor.cpu().numpy() if chr(c) != \"-\")\n",
    "                    outfile.write(f\"{current_header}\\n{trimmed_sequence}\\n\")\n",
    "\n",
    "                # Update current header\n",
    "                current_header = line.strip()\n",
    "                current_sequence = \"\"\n",
    "            else:\n",
    "                current_sequence += line.strip()\n",
    "\n",
    "        # Write the last sequence\n",
    "        if current_sequence:\n",
    "            # Convert the sequence to a tensor\n",
    "            sequence_tensor = torch.tensor([ord(c) for c in current_sequence], dtype=torch.int32, device='cuda')\n",
    "            # Trim and replace \"-\" with \"\"\n",
    "            trimmed_sequence_tensor = sequence_tensor[start_position - 1:end_position]\n",
    "            trimmed_sequence = ''.join(chr(c) for c in trimmed_sequence_tensor.cpu().numpy() if chr(c) != \"-\")\n",
    "            outfile.write(f\"{current_header}\\n{trimmed_sequence}\\n\")\n",
    "\n",
    "    print(f\"Trimmed sequences between positions {start_position} and {end_position} to '{output_fasta}'.\")\n",
    "\n",
    "def main():\n",
    "    input_fasta = \"your/path/msaCodon_1024_filtered.fasta\"\n",
    "    output_fasta = \"your/path/msaCodon_1024_trimmed_RBD.fasta\"\n",
    "    start_position = 35011\n",
    "    end_position = 36610\n",
    "\n",
    "    trim_and_write_sequences(input_fasta, output_fasta, start_position, end_position)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f1cad02",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Counting sequence statistics for trimmed RBD fasta file to find out if outliers exist\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "def generate_sequence_statistics(fasta_file):\n",
    "    sequence_lengths = []\n",
    "    total_sequences = 0\n",
    "    total_bases = 0\n",
    "\n",
    "    with open(fasta_file, \"r\") as handle:\n",
    "        for record in SeqIO.parse(handle, \"fasta\"):\n",
    "            total_sequences += 1\n",
    "            sequence_length = len(reco rd.seq)\n",
    "            sequence_lengths.append(sequence_length)\n",
    "            total_bases += sequence_length\n",
    "\n",
    "    # Calculate statistics\n",
    "    min_length = min(sequence_lengths)\n",
    "    max_length = max(sequence_lengths)\n",
    "    average_length = total_bases / total_sequences\n",
    "\n",
    "    # Print the statistics\n",
    "    print(\"Sequence Statistics:\")\n",
    "    print(f\"Total Sequences: {total_sequences}\")\n",
    "    print(f\"Total Bases: {total_bases}\")\n",
    "    print(f\"Minimum Length: {min_length} bases\")\n",
    "    print(f\"Maximum Length: {max_length} bases\")\n",
    "    print(f\"Average Length: {average_length:.2f} bases\")\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    fasta_file = \"your/path/msaCodon_1024_trimmed_RBD.fasta\"  # Replace with your FASTA file\n",
    "    generate_sequence_statistics(fasta_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "255f694a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate first quartile (Q1), third quartile (Q3), and the interquartile range (IQR) and \n",
    "# use the IQR to determine the lower and upper bounds for outlier detection and remove outlier readlengths.\n",
    "\n",
    "import numpy as np\n",
    "from collections import defaultdict\n",
    "\n",
    "def count_sequence_lengths(fasta_file):\n",
    "    sequence_lengths = []\n",
    "    sequences = []\n",
    "\n",
    "    with open(fasta_file, \"r\") as file:\n",
    "        sequence = \"\"\n",
    "        header = \"\"\n",
    "\n",
    "        for line in file:\n",
    "            if line.startswith(\">\"):\n",
    "                if sequence:\n",
    "                    sequence_lengths.append(len(sequence))\n",
    "                    sequences.append((header, sequence))\n",
    "                header = line.strip()\n",
    "                sequence = \"\"\n",
    "            else:\n",
    "                sequence += line.strip()\n",
    "\n",
    "        # Don't forget to count the last sequence\n",
    "        if sequence:\n",
    "            sequence_lengths.append(len(sequence))\n",
    "            sequences.append((header, sequence))\n",
    "\n",
    "    return sequence_lengths, sequences\n",
    "\n",
    "def calculate_iqr_outliers(sequence_lengths):\n",
    "    q1 = np.percentile(sequence_lengths, 25)\n",
    "    q3 = np.percentile(sequence_lengths, 75)\n",
    "    iqr = q3 - q1\n",
    "    lower_bound = q1 - 1.5 * iqr\n",
    "    upper_bound = q3 + 1.5 * iqr\n",
    "    return lower_bound, upper_bound\n",
    "\n",
    "def filter_outliers_and_write(sequences, lower_bound, upper_bound, output_file):\n",
    "    with open(output_file, \"w\") as outfile:\n",
    "        for header, sequence in sequences:\n",
    "            if lower_bound <= len(sequence) <= upper_bound:\n",
    "                outfile.write(f\"{header}\\n{sequence}\\n\")\n",
    "\n",
    "def main():\n",
    "    fasta_file = \"your/path/msaCodon_1024_trimmed_RBD.fasta\"\n",
    "    output_file = \"your/path/filtered_msaCodon_1024_trimmed_RBD.fasta\"\n",
    "\n",
    "    sequence_lengths, sequences = count_sequence_lengths(fasta_file)\n",
    "    lower_bound, upper_bound = calculate_iqr_outliers(sequence_lengths)\n",
    "    filter_outliers_and_write(sequences, lower_bound, upper_bound, output_file)\n",
    "\n",
    "    print(f\"Sequences between lengths {int(lower_bound)} and {int(upper_bound)} written to '{output_file}'.\")\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1075e6b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Python script to make a subset of the filtered_msaCodon_1024_trimmed_RBD.fasta without N and  \n",
    "## also make sure length of each sequence is multiple of 3 \n",
    "\n",
    "from Bio import SeqIO\n",
    "import random\n",
    "\n",
    "# Input FASTA file and output file\n",
    "input_fasta = \"your/path/filtered_msaCodon_1024_trimmed_RBD.fasta\"\n",
    "output_fasta = \"your/path/filtered_msaCodon_1024_trimmed_RBD_3mil.fasta\"\n",
    "\n",
    "# Read all sequences from the input FASTA file\n",
    "all_sequences = list(SeqIO.parse(input_fasta, \"fasta\"))\n",
    "\n",
    "# Randomly select 1 million sequences\n",
    "num_sequences_to_select = 3000000\n",
    "subset_sequences = random.sample(all_sequences, min(num_sequences_to_select, len(all_sequences)))\n",
    "\n",
    "# Filter sequences to contain only A, T, G, and C, and have lengths as multiples of 3\n",
    "valid_sequences = [seq for seq in subset_sequences if set(str(seq.seq)).issubset(\"ATGC\")]\n",
    "\n",
    "# Ensure each selected sequence length is a multiple of 3 or remove 1 or 2 nucleotides\n",
    "for i in range(len(valid_sequences)):\n",
    "    seq_len = len(valid_sequences[i])\n",
    "    if seq_len % 3 != 0:\n",
    "        trim_length = seq_len % 3\n",
    "        valid_sequences[i] = valid_sequences[i][:-trim_length]\n",
    "\n",
    "# Write the selected sequences to the output FASTA file\n",
    "with open(output_fasta, \"w\") as output_handle:\n",
    "    SeqIO.write(valid_sequences, output_handle, \"fasta\")\n",
    "\n",
    "print(f\"Randomly selected {num_sequences_to_select} valid sequences and saved to {output_fasta}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96f866d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Script to write out matching fasta files with Accession ID values in metadata file to csv files with Variant and Lineage values\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import pandas as pd\n",
    "import json\n",
    "from tqdm import tqdm\n",
    "from multiprocessing import Pool, cpu_count, Manager\n",
    "\n",
    "# Load your metadata DataFrame (df_metadata) and provide the correct path to your FASTA file and \"RBD.voc.json\" file.\n",
    "csv_file = \"your/path/metadata.tsv\"\n",
    "df_metadata = pd.read_csv(csv_file, sep='\\t', dtype=str, low_memory=False, encoding='latin-1')\n",
    "\n",
    "\n",
    "# Define the output CSV file\n",
    "output_csv_file = \"your/path/filtered_msaCodon_1024_trimmed_RBD_3mil.csv\"\n",
    "\n",
    "# Initialize a count for matches\n",
    "match_count = 0\n",
    "total_sequences = 0\n",
    "\n",
    "# Function to process each sequence\n",
    "def process_sequence(record):\n",
    "    global df_metadata, voc_data\n",
    "    accession_id_fasta = record.description.split(\"|\")[1]\n",
    "    sequence = str(record.seq)\n",
    "\n",
    "    # Check if the Accession ID is present in the metadata DataFrame\n",
    "    match_row = df_metadata[df_metadata['Accession ID'] == accession_id_fasta]\n",
    "    if not match_row.empty:\n",
    "        lineage = match_row['Pango lineage'].values[0]\n",
    "        variant = match_row['Variant'].values[0]\n",
    "\n",
    "        # Create a dictionary for the current record\n",
    "        output_data = {\n",
    "            'Accession ID': accession_id_fasta,\n",
    "            'Lineage': lineage,\n",
    "            'RBD nucleotide': sequence,\n",
    "            'Variant': variant\n",
    "        }\n",
    "        return output_data\n",
    "    return None\n",
    "\n",
    "# Write the header to the output CSV file\n",
    "with open(output_csv_file, 'w', newline='') as f:\n",
    "    writer = pd.DataFrame(columns=['Accession ID', 'Lineage', 'RBD nucleotide', 'Variant'])\n",
    "    writer.to_csv(f, index=False)\n",
    "\n",
    "# Estimate the total number of sequences for tqdm\n",
    "with open(\"your/path/filtered_msaCodon_1024_trimmed_RBD_3mil.fasta\", \"r\") as fasta_file:\n",
    "    total_sequences_estimated = sum(1 for line in fasta_file if line.startswith(\">\"))\n",
    "\n",
    "# Process sequences using multiprocessing Pool\n",
    "with Pool(processes = 32) as pool, tqdm(total=total_sequences_estimated, desc=\"Processing sequences\") as pbar:\n",
    "    for result in pool.imap(process_sequence, SeqIO.parse(\"your/path/filtered_msaCodon_1024_trimmed_RBD_3mil.fasta\", \"fasta\")):\n",
    "        pbar.update()\n",
    "        if result:\n",
    "            with open(output_csv_file, 'a', newline='') as f:\n",
    "                output_df = pd.DataFrame([result])\n",
    "                output_df.to_csv(f, mode='a', header=False, index=False)\n",
    "                match_count += 1\n",
    "\n",
    "# Print the results\n",
    "print(f\"Matches found: {match_count}\")\n",
    "print(f\"Total sequences: {total_sequences_estimated}\")\n",
    "print(f\"Output CSV file saved as {output_csv_file}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
